nanoFramework.CoreLibrary/System/SpanByte.cs

```diff
@@ -0,0 +1,172 @@
+// Portions Copyright (c) Microsoft Corporation.  All rights reserved.
+// See LICENSE file in the project root for full license information.
+//
+
+namespace System
+{
+    /// <summary>
+    /// Provides a type- and memory-safe representation of a contiguous region of arbitrary byte array
+    /// </summary>
+    public struct SpanByte
+    {
+        private byte[] _array;
+        private readonly int _start;
+        private readonly int _length;
+
+        /// <summary>
+        /// Creates a new System.SpanByte object over the entirety of a specified array.
+        /// </summary>
+        /// <param name="array">The array from which to create the System.Span object.</param>
+        public SpanByte(byte[] array)
+        {
+            _array = array;
+            _length = array != null ? array.Length : 0;
+            _start = 0;
+        }
+
+        /// <summary>
+        /// Creates a new System.SpanByte object that includes a specified number of elements
+        /// of an array starting at a specified index.
+        /// </summary>
+        /// <param name="array">The source array.</param>
+        /// <param name="start">The index of the first element to include in the new System.Span</param>
+        /// <param name="length">The number of elements to include in the new System.Span</param>
+        /// <exception cref="System.ArgumentOutOfRangeException">
+        /// array is null, but start or length is non-zero. -or- start is outside the bounds
+        /// of the array. -or- start and length exceeds the number of elements in the array.
+        /// </exception>
+        public SpanByte(byte[] array, int start, int length)
+        {
+            if (array != null)
+            {
+                if ((length > array.Length - start) || (start >= array.Length))
+                {
+                    throw new ArgumentOutOfRangeException($"Array length too small");
+                }
+            }
+            else
+            {
+                if ((start != 0) || (length != 0))
+                {
+                    throw new ArgumentOutOfRangeException($"Array is null but start and length are not 0");
+                }
+            }
+
+            _array = array;
+            _start = start;
+            _length = length;
+        }
+
+        /// <summary>
+        /// Gets the element at the specified zero-based index.
+        /// </summary>
+        /// <param name="index">The zero-based index of the element.</param>
+        /// <returns>The element at the specified index.</returns>
+        public byte this[int index]
+        {
+            get
+            {
+                return _array[_start + index];
+            }
+            set
+            {
+                _array[_start + index] = value;
+            }
+        }
+
+        /// <summary>
+        /// Returns an empty System.Span object.
+        /// </summary>
+        public static SpanByte Empty => new SpanByte();
```

Comment:
Looking at this one I think we should have a default constructor setting `_array = new byte[0];`.
Otherwise anything trying to access the array will throw a NULL exception.

Reply:
It's an ok behavior, if null, length and start will be 0